File: src/commands/approve.ts
// `specd approve <slug> [--json]` — the human/orchestrator approval primitive (SPEC §5.2, §7.6).
// Two jobs, dispatched on state: (1) clear a `midreq`-set awaiting-approval gate so work resumes;
// (2) advance the planning ratchet requirements → design → tasks → executing, but only once the
// gate for the artifact that phase produced is green. This is the inverse of `midreq` and the
// only sanctioned way to clear the gate or advance a planning phase (state.json stays CLI-owned).
import { requireSpecdRoot } from "../core/paths.js";
import { loadSpec, readArtifact, loadConfig } from "../core/specFiles.js";
import { withSpecLock } from "../core/lock.js";
import { saveState } from "../core/state.js";
import { PLANNING_ADVANCE, phaseReadiness, phaseForStatus } from "../core/phases.js";
import { acceptanceGaps } from "../core/render.js";
import { usageError, gateError } from "../core/exit.js";
import type { Args } from "../cli.js";

export function run(args: Args): number {
  const root = requireSpecdRoot();
  const slug = args.pos[0];
  if (!slug) throw usageError("usage: specd approve <slug> [--json]");
  const json = args.flags.json === true;
  // Lock across load→advance→save so a concurrent approve/task can't race the planning ratchet.
  return withSpecLock(root, slug, () => {
  const { state, doc } = loadSpec(root, slug);

  // Case 1: a midreq raised the gate — clear it, leave status/phase where execution paused.
  if (state.gate === "awaiting-approval") {
    state.gate = "none";
    saveState(root, slug, state);
    if (json) { console.log(JSON.stringify({ ok: true, action: "gate-cleared", status: state.status, phase: state.phase }, null, 2)); return 0; }
    console.log(`approve: gate cleared — resume at status '${state.status}' (phase ${state.phase}).`);
    return 0;
  }

  // Case 2: spec-level VERIFY → REFLECT. Every task is complete (status `verifying`); this is the
  // human acceptance of the spec-level verification. Advance to `complete`/reflect.
  if (state.status === "verifying") {
    // G5: when the acceptance gate is `required`, refuse while any requirement lacks a passing
    // criterion or any criterion is recorded as fail — the per-criterion proof must be complete.
    if (loadConfig(root).gates.acceptance === "required") {
      const { unmet, failed } = acceptanceGaps(state, readArtifact(root, slug, "requirements.md"));
      if (unmet.length || failed.length) {
        const problems = [
          ...unmet.map((n) => `requirement ${n}: no passing acceptance criterion`),
          ...failed.map((k) => `criterion ${k}: recorded as fail`),
        ];
        if (json) { console.log(JSON.stringify({ ok: false, action: "blocked", status: state.status, problems }, null, 2)); return 1; }
        for (const p of problems) console.error(`fail  ${p}`);
        console.error(`\n✗ cannot approve verification — ${problems.length} unmet acceptance criterion/criteria. Record with \`specd verify ${slug} --criterion <r>.<n> --status pass --evidence "..."\`.`);
        return 1;
      }
    }
    const from = state.status;
    state.status = "complete";
    state.phase = phaseForStatus("complete");
    saveState(root, slug, state);
    if (json) { console.log(JSON.stringify({ ok: true, action: "verified", from, status: state.status, phase: state.phase }, null, 2)); return 0; }
    console.log(`approve: verification accepted → status 'complete' (phase ${state.phase}).`);
    return 0;
  }

  // Case 3: advance the planning ratchet.
  const advance = PLANNING_ADVANCE[state.status];
  if (!advance) {
    throw gateError(`approve: nothing to approve — spec '${slug}' is '${state.status}'.`);
  }

  const problems = phaseReadiness(
    state.status,
    readArtifact(root, slug, "requirements.md"),
    readArtifact(root, slug, "design.md"),
    doc,
  );
  if (problems.length) {
    if (json) { console.log(JSON.stringify({ ok: false, action: "blocked", status: state.status, problems }, null, 2)); return 1; }
    for (const p of problems) console.error(`fail  ${p}`);
    console.error(`\n✗ cannot approve '${state.status}' — ${problems.length} gate violation(s). Fix and retry.`);
    return 1;
  }

  const from = state.status;
  state.status = advance.status;
  state.phase = advance.phase;
  saveState(root, slug, state);

  if (json) { console.log(JSON.stringify({ ok: true, action: "advanced", from, status: state.status, phase: state.phase }, null, 2)); return 0; }
  console.log(`approve: '${from}' approved → status '${state.status}' (phase ${state.phase}).`);
  return 0;
  });
}

File: internal/cmd/approve.go
package cmd

import (
	"fmt"

	"github.com/0xkhdr/specd/internal/cli"
	"github.com/0xkhdr/specd/internal/core"
)

// RunApprove implements `specd approve`. Under the spec lock it handles three
// cases in order: clearing an awaiting-approval gate, accepting a verifying
// spec as complete (enforcing the acceptance-criteria gate when configured),
// or advancing a planning-phase spec to the next status — printing the result
// as JSON or human-readable text.
//
//nolint:gocyclo // pre-existing complexity debt, out of scope for spec S3 — tracked for a future cleanup pass
func RunApprove(args cli.Args) int {
	root, slug, code, ok := requireRootAndSlug(args, "usage: specd approve <slug> [--json]")
	if !ok {
		return code
	}
	jsonOut := args.Bool("json")

	result, err := core.WithSpecLock[int](root, slug, func() (int, error) {
		loaded, err := core.LoadSpec(root, slug)
		if err != nil {
			return specdExit(err), err
		}
		state := loaded.State
		doc := loaded.Doc

		// Case 1: clear awaiting-approval gate.
		if state.Gate == core.GateAwaitingApproval {
			state.Gate = core.GateNone
			if err := core.SaveState(root, slug, state); err != nil {
				return specdExit(err), err
			}
			if jsonOut {
				if err := core.PrintJSON(map[string]interface{}{"ok": true, "action": "gate-cleared", "status": state.Status, "phase": state.Phase}); err != nil {
					return specdExit(err), err
				}
			} else {
				fmt.Printf("approve: gate cleared — resume at status '%s' (phase %s).\n", state.Status, state.Phase)
			}
			return core.ExitOK, nil
		}

		// Case 2: verifying → complete.
		if state.Status == core.StatusVerifying {
			cfg := core.LoadConfig(root)
			if cfg.Gates.Acceptance == "required" {
				gaps := core.GetAcceptanceGaps(state, core.ReadArtifact(root, slug, "requirements.md"))
				if len(gaps.Unmet) > 0 || len(gaps.Failed) > 0 {
					var problems []string
					for _, n := range gaps.Unmet {
						problems = append(problems, fmt.Sprintf("requirement %d: no passing acceptance criterion", n))
					}
					for _, k := range gaps.Failed {
						problems = append(problems, fmt.Sprintf("criterion %s: recorded as fail", k))
					}
					if jsonOut {
						if err := core.PrintJSON(map[string]interface{}{"ok": false, "action": "blocked", "status": state.Status, "problems": problems}); err != nil {
							return specdExit(err), err
						}
					} else {
						for _, p := range problems {
							errLine("fail  %s", p)
						}
						errLine("\n✗ cannot approve verification — %d unmet acceptance criterion/criteria. Record with `specd verify %s --criterion <r>.<n> --status pass --evidence \"...\"`.", len(problems), slug)
					}
					return core.ExitGate, nil
				}
			}
			// Prototype specs can never reach `complete` — they must be promoted
			// to full specs first (V5 prototype lifecycle, invariant 5).
			if state.Prototype != nil && state.Prototype.Status != core.PrototypePromoted {
				msg := fmt.Sprintf("cannot complete prototype spec '%s' — run `specd promote %s --evidence \"...\"` to convert it to a full spec first", slug, slug)
				if jsonOut {
					if err := core.PrintJSON(map[string]interface{}{"ok": false, "action": "blocked", "status": state.Status, "problems": []string{msg}}); err != nil {
						return specdExit(err), err
					}
				} else {
					errLine("✗ %s", msg)
				}
				return core.ExitGate, nil
			}
			// Review gate: when configured required, completion needs a fresh,
			// structurally-valid review_report.md whose verdict is `approve` (off for
			// migrated repos). Human approval stays final — this only enforces that the
			// review evidence exists and is current, never substitutes for it.
			if cfg.Review.Required {
				body, mod := core.ReadReviewReport(root, slug)
				res := core.EvaluateReviewGate(state, body, mod)
				if !res.OK {
					msg := "review gate: " + res.Problem
					if jsonOut {
						if err := core.PrintJSON(map[string]interface{}{"ok": false, "action": "blocked", "status": state.Status, "problems": []string{msg}}); err != nil {
							return specdExit(err), err
						}
					} else {
						errLine("✗ %s", msg)
					}
					return core.ExitGate, nil
				}
				state.Review = &core.ReviewRecord{Verdict: string(res.Verdict), Fresh: res.Fresh, Time: core.NowISO()}
			}
			// Eval gate: when configured `required`, completion needs at least one
			// passing recorded rubric run (config-on for new inits, off for
			// migrated repos — gate-fatigue mitigation).
			if cfg.Gates.Eval == "required" && !hasPassingEval(state) {
				msg := fmt.Sprintf("eval gate: no passing eval run recorded for '%s' — run `specd eval %s` (score ≥ minScore) before completing", slug, slug)
				if jsonOut {
					if err := core.PrintJSON(map[string]interface{}{"ok": false, "action": "blocked", "status": state.Status, "problems": []string{msg}}); err != nil {
						return specdExit(err), err
					}
				} else {
					errLine("✗ %s", msg)
				}
				return core.ExitGate, nil
			}
			from := state.Status
			state.Status = core.StatusComplete
			state.Phase = core.PhaseForStatus(core.StatusComplete)
			if err := core.SaveState(root, slug, state); err != nil {
				return specdExit(err), err
			}
			if jsonOut {
				if err := core.PrintJSON(map[string]interface{}{"ok": true, "action": "verified", "from": from, "status": state.Status, "phase": state.Phase}); err != nil {
					return specdExit(err), err
				}
			} else {
				fmt.Printf("approve: verification accepted → status 'complete' (phase %s).\n", state.Phase)
			}
			return core.ExitOK, nil
		}

		// Case 3: planning ratchet.
		advance, ok := core.PlanningAdvance[state.Status]
		if !ok {
			return specdExit(core.GateError(fmt.Sprintf("approve: nothing to approve — spec '%s' is '%s'.", slug, state.Status))), core.GateError("")
		}
		// Prototype specs skip the design/tasks planning gates (V5): the ratchet
		// still advances status but PhaseReadiness content checks are relaxed.
		var problems []string
		if state.Prototype == nil || state.Prototype.Status == core.PrototypePromoted {
			problems = core.PhaseReadiness(state.Status, core.ReadArtifact(root, slug, "requirements.md"), core.ReadArtifact(root, slug, "design.md"), doc)
		}
		if len(problems) > 0 {
			if jsonOut {
				if err := core.PrintJSON(map[string]interface{}{"ok": false, "action": "blocked", "status": state.Status, "problems": problems}); err != nil {
					return specdExit(err), err
				}
			} else {
				for _, p := range problems {
					errLine("fail  %s", p)
				}
				errLine("\n✗ cannot approve '%s' — %d gate violation(s). Fix and retry.", state.Status, len(problems))
			}
			return core.ExitGate, nil
		}
		from := state.Status
		state.Status = advance.Status
		state.Phase = advance.Phase
		if err := core.SaveState(root, slug, state); err != nil {
			return specdExit(err), err
		}
		if jsonOut {
			if err := core.PrintJSON(map[string]interface{}{"ok": true, "action": "advanced", "from": from, "status": state.Status, "phase": state.Phase}); err != nil {
				return specdExit(err), err
			}
		} else {
			fmt.Printf("approve: '%s' approved → status '%s' (phase %s).\n", from, state.Status, state.Phase)
		}
		return core.ExitOK, nil
	})
	if err != nil {
		return specdExit(err)
	}
	return result
}

// hasPassingEval reports whether any recorded eval suite passed its minScore.
// It reads only recorded state — deterministic, no re-run of the rubric.
func hasPassingEval(state *core.State) bool {
	for _, e := range state.Evals {
		if e.Pass {
			return true
		}
	}
	return false
}
